lib/jsdom/living/helpers/html-element-constructor.js

```diff
@@ -0,0 +1,162 @@
+"use strict";
+
+const DOMException = require("domexception");
+
+const { HTML_NS } = require("./namespaces");
+const { CUSTOM_ELEMENT_STATE } = require("./custom-elements");
+
+const { createElement, INTERFACE_TAG_MAPPING } = require("../create-element");
+const { implForWrapper, wrapperForImpl } = require("../generated/utils");
+
+// https://html.spec.whatwg.org/multipage/custom-elements.html#concept-already-constructed-marker
+const ALREADY_CONSTRUCTED_MARKER = Symbol("already-constructed-marker");
+
+const CONSTRUCTORS_TO_PATCH = [
+  "HTMLAnchorElement",
+  "HTMLAreaElement",
+  "HTMLAudioElement",
+  "HTMLBaseElement",
+  "HTMLBodyElement",
+  "HTMLBRElement",
+  "HTMLButtonElement",
+  "HTMLCanvasElement",
+  "HTMLDataElement",
+  "HTMLDataListElement",
+  "HTMLDetailsElement",
+  "HTMLDialogElement",
+  "HTMLDirectoryElement",
+  "HTMLDivElement",
+  "HTMLDListElement",
+  "HTMLElement",
+  "HTMLEmbedElement",
+  "HTMLFieldSetElement",
+  "HTMLFontElement",
+  "HTMLFormElement",
+  "HTMLFrameElement",
+  "HTMLFrameSetElement",
+  "HTMLHeadElement",
+  "HTMLHeadingElement",
+  "HTMLHRElement",
+  "HTMLHtmlElement",
+  "HTMLIFrameElement",
+  "HTMLImageElement",
+  "HTMLInputElement",
+  "HTMLLabelElement",
+  "HTMLLegendElement",
+  "HTMLLIElement",
+  "HTMLLinkElement",
+  "HTMLMapElement",
+  "HTMLMarqueeElement",
+  "HTMLMenuElement",
+  "HTMLMetaElement",
+  "HTMLMeterElement",
+  "HTMLModElement",
+  "HTMLObjectElement",
+  "HTMLOListElement",
+  "HTMLOptGroupElement",
+  "HTMLOptionElement",
+  "HTMLOutputElement",
+  "HTMLParagraphElement",
+  "HTMLParamElement",
+  "HTMLPictureElement",
+  "HTMLPreElement",
+  "HTMLProgressElement",
+  "HTMLQuoteElement",
+  "HTMLScriptElement",
+  "HTMLSelectElement",
+  "HTMLSlotElement",
+  "HTMLSourceElement",
+  "HTMLSpanElement",
+  "HTMLStyleElement",
+  "HTMLTableCaptionElement",
+  "HTMLTableCellElement",
+  "HTMLTableColElement",
+  "HTMLTableElement",
+  "HTMLTableRowElement",
+  "HTMLTableSectionElement",
+  "HTMLTemplateElement",
+  "HTMLTextAreaElement",
+  "HTMLTimeElement",
+  "HTMLTitleElement",
+  "HTMLTrackElement",
+  "HTMLUListElement",
+  "HTMLVideoElement",
+];
+
+function patchHTMLConstructor(window) {
```

Comment:
This implementation of `[HTMLConstructor]` patches the interface constructor and keep in place the interface prototype. This approach avoids paying the performance overhead of creating a brand new interface per window. Is there something that can be done at webidl2js level to avoid hardcoding a list of interfaces to patch?

Reply:
This is the hard one. See below for my thoughts on webidl2js plugins and interfacing. I think a `processHTMLConstructor` option might be the best bet here, so that webidl2js calls you every time it sees HTMLConstructor, and asks you to provide the text it wants to generate.

Then, you can reuse the architecture of https://github.com/jsdom/webidl2js#webidl2jsfactory so that it generates constructors per-Window in the same way. But, right now we just manually know that there are 4 interfaces which need to be set up in Window's constructor. (See e.g. the `createAbortController` call.) It'd be best to generalize this, so that webidl2js spits out a file we can require(), which exports a function we can use like `installPerWindowConstructors(window)`. That would then take care of both [Webidl2jsFactory] and [HTMLConstructor].